wandb/sync/sync.py

```diff
@@ -140,26 +149,58 @@ def _send_tensorboard(self, tb_root, tb_logdirs, send_manager):
         )
         print("Syncing: %s ..." % url)
         sys.stdout.flush()
+        # this sucks, but if we want to add in the Handler specifically for tb case
+        # we need to remake the SendManager and Backend Sender
+        record_q = queue.Queue()
+        sender_record_q = queue.Queue()
+        new_interface = interface.BackendSender(record_q)
+        send_manager = sender.SendManager(
+            send_manager._settings, sender_record_q, queue.Queue(), new_interface
+        )
         record = send_manager._interface._make_record(run=proto_run)
-        send_manager.send(record)
         settings = wandb.Settings(
             root_dir=TMPDIR.name,
             run_id=proto_run.run_id,
             _start_datetime=datetime.datetime.now(),
             _start_time=time.time(),
         )
-        watcher = tb_watcher.TBWatcher(
-            settings, proto_run, send_manager._interface, True
+
+        handle_manager = handler.HandleManager(
+            settings, record_q, None, False, sender_record_q, None, new_interface
         )
+
+        mkdir_exists_ok(settings.files_dir)
+        send_manager.send_run(record, file_dir=settings.files_dir)
+        watcher = tb_watcher.TBWatcher(settings, proto_run, new_interface, True)
+
         for tb in tb_logdirs:
+            print("Adding watcher to event file", tb)
             watcher.add(tb, True, tb_root)
             sys.stdout.flush()
         watcher.finish()
+
         # send all of our records like a boss
-        while not send_manager._interface.record_q.empty():
-            data = send_manager._interface.record_q.get(block=True)
+        progress_step = 0
+        spinner_states = ["-", "\\", "|", "/"]
+        line = " Uploading data to wandb\r"
+        while not handle_manager._record_q.empty():
+
+            data = handle_manager._record_q.get(block=True)
+            handle_manager.handle(data)
+            if not send_manager._record_q.empty():
+                data = send_manager._record_q.get(block=True)
+                send_manager.send(data)
```

Comment:
Instead of calling this private attributes, let's add simpler helper methods to handle_manager and send_manager that do this.  We might be able to just make the managers iterators, so this loop could look like:

```
for data in handle_manager:
    handle_manager.handle(data)
    if len(send_manager) > 0:
        data = next(send_manager)
        send_manager.send(data)
```

It's also a bit awkward that we're only sending one record at a time in the handler thread.  We could likely just loop over everything in the send_manager record q.

Reply:
yeah, there's no reason to send one at a time. That `if` should be a `while`, to make it less awkward. I'll make the managers iterators as well.